org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -1225,6 +1285,20 @@ private void addRegistration(Registration reg, Runnable unregistrationHandler) {
 		}
 	}
 
+	synchronized void disableWatchedFiles() {
+		if (watchedFilesListenerInstalled && !fileSystemWatcherManager.hasFilePatterns()) {
```

Comment:
I wonder if we can remove the field `watchedFilesListenerInstalled`. From what I see in add and remove resource change listener, it is safe to remove if it has not been added, and it is safe to add the same listener twice, just a bit more expensive because it needs to go over the list of all listeners. Given that this is removal/addition does not happen frequently, I would do away with the class field.